src/transport/fusedev/fuse_t_session.rs

```diff
@@ -0,0 +1,638 @@
+        let result = file_lock.lock();
+        let fd = self.file.as_raw_fd();
+        let size = size_of::<InHeader>();
+        // read header
+        self.read(size, 0)?;
+        let in_header = InHeader::from_slice(&self.buf[0..size]);
+        let header_len = in_header.unwrap().len as usize;
+        let should_read_size = header_len - size;
+        if should_read_size > 0 {
+            self.read(should_read_size, size)?;
+        }
+        drop(result);
+
+        let buf = unsafe { std::slice::from_raw_parts_mut(self.buf.as_mut_ptr(), self.buf.len()) };
+        // Reader::new() and Writer::new() should always return success.
+        let reader = Reader::from_fuse_buffer(FuseBuf::new(&mut self.buf[..header_len])).unwrap();
+        let writer = FuseDevWriter::new(fd, buf).unwrap();
+        Ok(Some((reader, writer)))
+    }
+}
+
+fn fuse_kern_mount(
+    mountpoint: &Path,
+    fsname: &str,
+    subtype: &str,
+    rd_only: bool,
+) -> Result<(File, File)> {
+    unsafe { signal(Signal::SIGCHLD, SigHandler::SigDfl) }
+        .map_err(|e| SessionFailure(format!("fail to reset SIGCHLD handler{:?}", e)))?;
+
+    let (fd0, fd1) = socketpair(
+        AddressFamily::Unix,
+        SockType::Stream,
+        None,
+        SockFlag::empty(),
+    )
+    .map_err(|e| SessionFailure(format!("create socket failed {:?}", e)))?;
+
+    setsockopt(fd0, SndBuf, &FS_SND_SIZE)
+        .map_err(|e| SessionFailure(format!("set fd0 socket snd size {:?}", e)))?;
+    setsockopt(fd0, RcvBuf, &FS_SND_SIZE)
+        .map_err(|e| SessionFailure(format!("set fd0 socket rcv size {:?}", e)))?;
+    setsockopt(fd1, SndBuf, &FS_SND_SIZE)
+        .map_err(|e| SessionFailure(format!("set fd1 socket snd size {:?}", e)))?;
+    setsockopt(fd1, RcvBuf, &FS_SND_SIZE)
+        .map_err(|e| SessionFailure(format!("set fd1 socket rcv size {:?}", e)))?;
+
+    let (mon_fd0, mon_fd1) = socketpair(
+        AddressFamily::Unix,
+        SockType::Stream,
+        None,
+        SockFlag::empty(),
+    )
+    .map_err(|e| SessionFailure(format!("create mon socket failed {:?}", e)))?;
+
+    let res;
+    unsafe {
+        res = fork().map_err(|e| SessionFailure(format!("fork mount_macfuse failed {:?}", e)))?;
+    }
+
+    match res {
+        ForkResult::Parent { .. } => {
+            close(fd0).map_err(|e| SessionFailure(format!("parent close fd0 failed {:?}", e)))?;
+            close(mon_fd0)
+                .map_err(|e| SessionFailure(format!("parent close mon fd0 failed {:?}", e)))?;
+            unsafe { Ok((File::from_raw_fd(fd1), File::from_raw_fd(mon_fd1))) }
+        }
+        ForkResult::Child => {
+            close(fd1).map_err(|e| SessionFailure(format!("child close fd1 failed {:?}", e)))?;
+            close(mon_fd1)
+                .map_err(|e| SessionFailure(format!("child close mon fd1 failed {:?}", e)))?;
+
+            let mut daemon_path: Vec<u8> = Vec::with_capacity(PROC_PIDPATHINFO_MAXSIZE as usize);
+            unsafe {
```

Comment:
please limit the scope of `unsafe` to just `proc_pidpath`. IIUC, it is the only one that requires `unsafe`, right?

Reply:
And `daemon_path.set_len` is unsafe too.